src/dune_engine/rule.ml

```diff
@@ -90,6 +91,13 @@ let add_sandbox_config :
 
 let make ?(sandbox = Sandbox_config.default) ?(mode = Mode.Standard) ~context
     ?(info = Info.Internal) ~targets action =
+  let sandbox =
+    match mode with
+    | Patch_back_source_tree ->
+      Sandbox_config.inter sandbox
```

Comment:
If the resulting set of allowed sandboxing modes is going to become empty because of this `inter`, we aren't going to get a great error message. I suggest adding a test to illustrate this.

Reply:
Indeed, I added a test for this and we get:

```
  $ dune build
  File "dune", line 1, characters 0-119:
  1 | (rule
  2 |  (mode patch-back-source-tree)
  3 |  (deps (sandbox none))
  4 |  (alias default)
  5 |  (action (system "echo 'Hello, world!'")))
  Error: This rule forbids all sandboxing modes (but it also requires
  sandboxing)
  [1]
```

which is not perfect, but not terrible either. It should be relatively easy to get a better error message at parsing time. However, it's a bit awkward to do in Dune as the sandbox config is specified in the deps field though. So I propose to do nothing here and inside JS we'll add the check at parsing time. Once we change the way the sandbox config is specified in Dune we can add a check at parsing time.

In any case, I added a comment about it.

Reply:
Thanks, sounds good! I think one way to address this problem in Dune in future is to change `Sandbox_mode` to carry the provenance of each restriction. I was thinking about it a bit when working on directory targets (which also insist on sandboxing) and it seemed doable.